docs/RFCS/20220120_row_level_ttl.md

```diff
@@ -0,0 +1,364 @@
+following on the deletion job itself:
+
+Option | Description
+--- | ---
+`expire_after` | When a TTL would expire. Accepts any interval. Defaults to ''30 days''. Minimum of `'5 minutes'`.
+`ttl_select_batch_size` | How many rows to fetch from the range that have expired at a given time. Defaults to 500. Must be at least `1`.
+`ttl_delete_batch_size` | How many rows to delete at a time. Defaults to 100. Must be at least `1`.
+`ttl_range_concurrency` | How many concurrent ranges are being worked on at a time. Defaults to `cpu_core_count`. Must be at least `1`.
+`ttl_admission_control_priority` | Priority of the admission control to use when deleting. Set to a high amount if deletion traffic is the priority.
+`ttl_select_as_of_system_time` | AS OF SYSTEM TIME to use for the SELECT clause. Defaults to `-30s`. Proposed as a hidden option, as I'm not sure anyone will actually need it.
+`ttl_maximum_rows_deleted_per_second` | Maximum number of rows to be deleted per second (acts as the rate limit). Defaults to 0 (signifying none).
+`ttl_pause` | Pauses the TTL job from executing. Existing jobs will need to be paused using CANCEL JOB.
+`ttl_job_cron` | Frequency the job runs, specified using the CRON syntax.
+
+#### A note on delete_batch_size
+Deleting rows on tables with secondary indexes on a table lays more intents,
+hence causing higher contention. In this case, there is an argument to delete
+less rows at a time (or indeed, just one row at a time!) to avoid this. However,
+testing showed deletes could never catch up in this case and realistically a
+very small number of rows can be deleted in this way. "100" was set as the
+balance for now.
+
+## Filtering out Expired Rows
+Rows that have expired their TTL can be optionally removed from all SQL query
+results. However, this is not planned to be implemented for the first release of
+TTL.
+
+#### User override for the `crdb_internal_expiration` column
+In an ideal world, the user can simply update the `crdb_internal_expiration`
+value to "override" the TTL. However, with our current defaults, this value
+would get overridden if the row is updated due to the `ON UPDATE` clause. The
+`ON UPDATE` should ideally be:
+
+```sql
+CASE WHEN crdb_internal_expiration IS NULL
+-- NULL means the row should never be TTL'd
+THEN NULL
+-- otherwise, take whatever's larger: the user set crdb_internal_expiration
+-- value or an update TTL based on the current_timestamp.
+ELSE max(crdb_internal_expiration, current_timestamp() + 'table ttl value')
+```
+
+However, due to the limitation that `ON UPDATE` cannot reference a table
+expression, this cannot yet be implemented. This can change when we implement
+[triggers](https://github.com/cockroachdb/cockroach/issues/28296).
+
+A workaround may be planned after the first iteration, but it is not deemed
+a blocking feature for the first iteration of TTL.
+
+## Foreign Keys
+To avoid additional complexity in the initial implementation, foreign keys to or
```

Comment:
Are there databases out there that support TTL and foreign keys together?

Reply:
(see reviewable for comments on these)